# ??????????????????
# ???????????? read.csv ??????
# data <- read.csv("hap.txt")
file_path <- "hap/phe.txt"  # ???????????????????????????
data <- read.table(file_path, header = FALSE, sep = "\t")  # ????????????????????????????????????"\t"????????????????????????????????????","??????
# ?????????????????????
head(data)
colnames(data) <- c("sample", "Value")  # ????????????????????????
# ????????????????????????????????????????????????

# ???phe??????????????????????????????value???
sample_values <- data$Value  # ??????phe?????????value???????????????Value
sample_names <- data$sample
# ???????????????????????????
head(data)

genename <- "??????7"
chromosome <- "16"
start_position <- 10322777


#11322777  7138340  7238340 7538340 8322777 9322777
end_position <- 11322777

chromosome1 <- "contig16"

num <- 0
# ??????????????????VCF??????
subset_vcf <- vcf[seqnames(vcf) == chromosome1 &
                                  start(vcf) >= start_position &
                                  end(vcf) <= end_position]
library(dplyr)

# ???????????????????????????????????????
genotype_map <- c('0|0' = 1, '0|1' = 2, '1|0' = 2, '1|1' = 3)

# ????????????????????????????????????
sample_indexes <- match(sample_names, colnames(geno(subset_vcf)$GT))

result_list <- lapply(seq_along(start(subset_vcf)), function(variant) {
  variant_type <- ifelse(
    abs(nchar(ref(subset_vcf)[[variant]]) - nchar(alt(subset_vcf)[[variant]])) > 0,
    'indel',
    'snp'
  )
 
  print(as.character(start(subset_vcf)[variant])) 
  new_row <- data.frame(
    Type = variant_type,
    Variant_Position = as.character(start(subset_vcf)[variant]),
    stringsAsFactors = FALSE
    
  )
 
  genotypes <- geno(subset_vcf)$GT[variant, sample_indexes]
  sample_values <- data[data$sample %in% sample_names, "Value"]
  
  count_indexes <- (sample_values - 1) * 3 + genotype_map[genotypes]
  
  counts <- matrix(0, nrow = 1, ncol = 9)
  unique_counts <- unique(count_indexes)
  counts[1, unique_counts] <- table(count_indexes)[as.character(unique_counts)]
  
  colnames(counts) <- paste0("Count_", 1:9)
  new_row <- cbind(new_row, counts)
  return(new_row)
 
})
final_result <- bind_rows(result_list)

# gene_count_data <- do.call(rbind, result_list)
print(head(final_result))
library(openxlsx)

write.xlsx(final_result, file = paste0(genename, ".xlsx"), rowNames = FALSE)
